### PR TITLE
Support for uploading deb source packages to packagecloud

### PR DESCRIPTION
**Pull request by j-rivero**
The PR adds support to upload source packages (.dsc) to packagecloud. According to the [packagecloud documentation](https://packagecloud.io/docs) the upload of Debian source packages is supported via the same `push` command applied to the other types of packages. ROS repositories will require source packages for new uploads in the near future, the PR should help  preparing colcon packagecloud for this situation.

I run the changes on `colcon-core` and see the push command in the dry run:
```bash
...

== Uploading artifacts to 'packagecloud' ...
-- Skip uploading package to packagecloud.io, pass --upload to actually upload artifacts

Existing packagecloud.io package differs from the artifact 'python3-colcon-core_0.6.1-1_all.deb'

$ package_cloud push dirk-thomas/colcon/ubuntu/xenial deb_dist/python3-colcon-core_0.6.1-1_all.deb
$ package_cloud push dirk-thomas/colcon/ubuntu/xenial deb_dist/python3-colcon-core_0.6.1-1.dsc

Existing packagecloud.io package differs from the artifact 'python3-colcon-core_0.6.1-1_all.deb'

$ package_cloud push dirk-thomas/colcon/ubuntu/bionic deb_dist/python3-colcon-core_0.6.1-1_all.deb
$ package_cloud push dirk-thomas/colcon/ubuntu/bionic deb_dist/python3-colcon-core_0.6.1-1.dsc

Existing packagecloud.io package differs from the artifact 'python3-colcon-core_0.6.1-1_all.deb'

$ package_cloud push dirk-thomas/colcon/ubuntu/focal deb_dist/python3-colcon-core_0.6.1-1_all.deb
$ package_cloud push dirk-thomas/colcon/ubuntu/focal deb_dist/python3-colcon-core_0.6.1-1.dsc

Existing packagecloud.io package differs from the artifact 'python3-colcon-core_0.6.1-1_all.deb'

$ package_cloud push dirk-thomas/colcon/debian/stretch deb_dist/python3-colcon-core_0.6.1-1_all.deb
$ package_cloud push dirk-thomas/colcon/debian/stretch deb_dist/python3-colcon-core_0.6.1-1.dsc

Existing packagecloud.io package differs from the artifact 'python3-colcon-core_0.6.1-1_all.deb'

$ package_cloud push dirk-thomas/colcon/debian/buster deb_dist/python3-colcon-core_0.6.1-1_all.deb
$ package_cloud push dirk-thomas/colcon/debian/buster deb_dist/python3-colcon-core_0.6.1-1.dsc
```

### CODE REVIEW

**Review by nuclearsandwich — APPROVED** [review]

File: publish_python/upload/packagecloud.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ def upload_packagecloud(*, artifacts, upload, config):
         destination = f'{repository}{version}'
         for artifact in artifacts:
             ext = os.path.splitext(artifact)[1]
-            if ext not in ('.deb', '.whl'):
+            if ext not in ('.deb', '.whl', '.dsc'):
```
Comment (dirk-thomas) [inline publish_python/upload/packagecloud.py]: Nitpick: can you please insert the new value in alphabetical order?

**Review by dirk-thomas — COMMENTED** [review]
> Existing packagecloud.io package differs from the artifact 'python3-colcon-core_0.6.1-1_all.deb'

I am wondering which version / commit hash of `colcon-core` you tried this dry run with? If you use the same tagged commit this message shouldn't appear if your environment is setup correctly.

File: publish_python/upload/packagecloud.py (hunk 1 above)
Comment (j-rivero) [reply publish_python/upload/packagecloud.py]: sure, 4add04c

**Comment by j-rivero** [comment]
> I am wondering which version / commit hash of `colcon-core` you tried this dry run with? If you use the same tagged commit this message shouldn't appear if your environment is setup correctly.

I did not pay attention since I was only looking for the `.dsc` files to appear. I run a new test from the exact tag and messages are still there. Just run a quick comparison of debians and the only difference is the name in of the Maintainer in the package metadata:
```
colcon-core/deb_dist on ❯ debdiff python3-colcon-core_0.6.1-1_all.deb orig-python3-colcon-core_0.6.1-1_all.deb 
File lists identical (after any substitutions)

Control files: lines which differ (wdiff format)
------------------------------------------------
[-Homepage: https://colcon.readthedocs.io-]
Maintainer: [-Jose Luis Rivero <jrivero@osrfoundation.org>-] {+Dirk Thomas <web@dirk-thomas.net>+}
```

I think that is fine.

**Comment by dirk-thomas** [comment]
Thanks for the patch. Merging this since the patch itself to push `.dsc` files (as well as referenced archives within) for future releases looks good.

Regarding getting source packages released for existing packages: I think we should avoid replacing existing Debian packages (`*_all.deb`). We can either decide to only apply this for future releases or only upload the source packages for existing packages (even though their content might not patch what was used to create the binary package - e.g. due to a different `stdeb` version or the local user running the command).

**Comment by j-rivero** [comment]
>  I think we should avoid replacing existing Debian packages (`*_all.deb`).

+1

>  I think we should avoid replacing existing Debian packages (`*_all.deb`). We can either decide to only apply this for future releases or only upload the source packages for existing packages (even though their content might not patch what was used to create the binary package - e.g. due to a different `stdeb` version or the local user running the command).

Current work is to provide source packages for all existing packages in bootstrap ROS repositories. The option of leaving the future releases the work of uploading source packages would delay the goal without date. I'm in favor of uploading only the source packages missing, I can patch this tool to do it, not sure if there is any chance to block current packages to be replaced in packagecloud to avoid any possible error.

**Comment by dirk-thomas** [comment]
> I'm in favor of uploading only the source packages missing, I can patch this tool to do it, not sure if there is any chance to block current packages to be replaced in packagecloud to avoid any possible error.

If you could put together a patch, or script or step-by-step instructions how to generate source packages (making sure that they are exactly matching the version of the already released Debian packages) and upload only those, I am happy to do the publishing.

**Comment by j-rivero** [comment]
> If you could put together a patch, or script or step-by-step instructions how to generate source packages (making sure that they are exactly matching the version of the already released Debian packages) and upload only those, I am happy to do the publishing.

Thanks Dirk. We were able to populate the ros_bootstrap repository with source packages for the current versions, new ones should use the patch in this PR. No action needs to be done.,

**Pull request merged**